### ISSUE TITLE
Options field is not updated when calling Devices[Unit].Update in a python plugin

### ISSUE THREAD

**Issue by pipiche38** [domoticz/domoticz#2733]
I'm creating devices with the following statement:

`Domoticz.Device(DeviceID=str(DeviceID),Name=str(t) + "-" + str(DeviceID) + "-" + str(Ep), Unit=FreeUnit(self, Devices), TypeName="kWh", Options={"Zigate":str(self.ListOfDevices[DeviceID]), "TypeName":t}).Create()`

But when I update the device with the following statement, the Options field is not updated in the DB
`Devices[Unit].Update(nValue=int(nValue), sValue=str(sValue), Options=str(Options), SignalLevel=int(rssi), BatteryLevel=int(BatteryLvl) , Color = Color_)`

**Comment by pipiche38** [comment]
Coming back on this issue where I'm stuck. Here is the Options parameter I'm passing to Update call in my python plugin and which is not updated to the DB.

It looks like the test : `if (pOptionsDict && PyDict_Check(pOptionsDict))` in line 947 of hardware/plugin/PythonObjects.cpp is false. So either my object is not a Python Dict, which I would be surprise as you can see here after!

I tried to add some debuging there, but I didn't find a way to give pOptionsDict to log.Log !

`
Options = {'ClusterType': 'Plug', 'Zigate': {'Ep': {'01': {'0000': {}, '0006': '00', '000c': {}, 'Type': 'Plug'}, '02': {'000c': '0.0', 'Type': 'Power/Meter'}}, 'Status': 'inDB', 'Heartbeat': '16', 'RIA': '7', 'RSSI': 51, 'Battery': {}, 'Model': 'lumi.plug', 'MacCapa': '8e', 'IEEE': '00158d00024d873d', 'Type': 'Plug/PowerMeter', 'ProfileID': {}, 'ZDeviceID': {}, 'DomoID': '1cb6', 'SQN': {}}}`

**Comment by pipiche38** [comment]
I have been able to put SetDeviceOptions in debug and here are the log that I'm getting:


`Sep 28 17:34:32 pi3 domoticz[3028]: 2018-09-28 17:34:32.654  (Zigate) UpdateDevice_v2 Option before Update = {'ClusterType': 'LvlControl', 'Zigate': {'ZDeviceID': '0200', 'Heartbeat': 0, 'Type': {}, 'IEEE': '20918a0000038f65', 'Ep': {'01': {'0015': {}, '0008': {}, '0100': {}, '0004': {}, '0005': {}, '0006': {}, 'fc20': {}, '0000': {}, '000a': {}, 'fc21': {}, '0003': {}}}, 'Model': 'shutter.Profalux', 'Battery': {}, 'MacCapa': '8e', 'Status': 'inDB', 'RIA': '2', 'ProfileID': '0104', 'DomoID': 'e86c', 'RSSI': 66, 'SQN': {}}}


Here is Options before call Update

`
Sep 28 17:43:44 pi3 domoticz[3028]: 2018-09-28 17:43:44.390  (Zigate) UpdateDevice_v2 Option before Update = {'ClusterType': 'LvlControl', 'Zigate': {'ZDeviceID': '0200', 'Heartbeat': 0, 'Type': {}, 'IEEE': '20918a0000038f65', 'Ep': {'01': {'0015': {}, '0008': {}, '0100': {}, '0004': {}, '0005': {}, '0006': {}, 'fc20': {}, '0000': {}, '000a': {}, 'fc21': {}, '0003': {}}}, 'Model': 'shutter.Profalux', 'Battery': {}, 'MacCapa': '8e', 'Status': 'inDB', 'RIA': '0', 'ProfileID': '0104', 'DomoID': 'e86c', 'RSSI': 57}}`

`Devices[Unit].Update(nValue=int(nValue), sValue=str(sValue), Options=Options)
`
Here after we see reaching SetDeviceOptions in SQLHelper

```

Sep 28 17:43:44 pi3 domoticz[3028]: 2018-09-28 17:43:44.416  Status: DEBUG : reaching SetDeviceOptions
Sep 28 17:43:44 pi3 domoticz[3028]: 2018-09-28 17:43:44.416  Status: DEBUG : setting options 'ClusterType:THZsQ29udHJvbA==;Zigate:eydaRGV2aWNlSUQnOiAnMDIwMCcsICdIZWFydGJlYXQnOiAnMScsICdUeXBlJzoge30sICdJRUVFJzogJzIwOTE4YTAwMDAwMzhmNjUnLCAnRXAnOiB7JzAxJzogeycwMDE1Jzoge30sICcwMDA4Jzoge30sICcwMTAwJzoge30sICcwMDA0Jzoge30sICcwMDA1Jzoge30sICcwMDA2Jzoge30sICdmYzIwJzoge30sICcwMDAwJzoge30sICcwMDBhJzoge30sICdmYzIxJzoge30sICcwMDAzJzoge319fSwgJ01vZGVsJzogJ3NodXR0ZXIuUHJvZmFsdXgnLCAnQmF0dGVyeSc6IHt9LCAnTWFjQ2FwYSc6ICc4ZScsICdTdGF0dXMnOiAnaW5EQicsICdSSUEnOiAnMCcsICdQcm9maWxlSUQnOiAnMDEwNCd9' on device 13
```
Here after is Option after the Update. It is not matching what was in Options before

`
Sep 28 17:43:44 pi3 domoticz[3028]: 2018-09-28 17:43:44.418  (Zigate) UpdateDevice_v2 Option after Update = {'ClusterType': 'LvlControl', 'Zigate': "{'ZDeviceID': '0200', 'Heartbeat': '1', 'Type': {}, 'IEEE': '20918a0000038f65', 'Ep': {'01': {'0015': {}, '0008': {}, '0100': {}, '0004': {}, '0005': {}, '0006': {}, 'fc20': {}, '0000': {}, '000a': {}, 'fc21': {}, '0003': {}}}, 'Model': 'shutter.Profalux', 'Battery': {}, 'MacCapa': '8e', 'Status': 'inDB', 'RIA': '0', 'ProfileID': '0104'}"}`

**Comment by pipiche38** [comment]
Here is a short plugin.py to reproduce the problem.

A device is created in the onStart (if not yet created).

Then in onHearbeat, we display the 
(1) Display Devices[1].Options Look at Heartbeat -- It is 0
(2) Display Options after increment of hearbeat

In the next cyce, you'll see that Hearbeat is not incremented !!!!!!!!!!!!
[plugin.py.zip](https://github.com/domoticz/domoticz/files/2429628/plugin.py.zip)

**Comment by gizmocuz** [comment]
You shall not post on the forum, post here, and send a private message

**Comment by pipiche38** [comment]
Already send you an MP

Envoyé de mon iPhone

**Comment by pipiche38** [comment]
Problem fix with recommendation from @Dnpwwo on the forum .

Still I think that the Plugin Framework must complain and not do anything without any error messages in the log.

**Comment by pipiche38** [comment]
Post by Dnpwwo » Saturday 29 September 2018 15:04

I suspect your problem is because during the create you correctly pass in a dictionary with string leys and string values but in the update you pass in a nested dictionary rather than a string.

It seemed to work if I changed line 78 to be: 

`Options['Zigate']=str({'Ep': {'01': {'0000': {}, '0003': {}, '0004': {}, '0005': {}, '0006': {}, '0008': {}, '0300': {}, '0b05': {}, '1000': {}}}, 'Status': 'inDB', 'Heartbeat': str(self.hb), 'RIA': '0', 'RSSI': {}, 'Battery': {}, 'Model': 'Ampoule.LED1624G9.Tradfri', 'MacCapa': '8e', 'IEEE': '90fd9ffffee77f7e', 'Type': {}, 'ProfileID': 'c05e', 'ZDeviceID': '0200', 'DomoID': '3145', 'SQN': {}})`

**Comment by pipiche38** [comment]
Fix in fd312856a12cbba86142b2751d4e197e83788cac
